src/QsCompiler/Core/SymbolTable.fs

```diff
@@ -1295,7 +1295,10 @@ and NamespaceManager
             let typeDiagnostics = this.CacheTypeResolution()
             let callableDiagnostics = this.CacheCallableResolutions()
             this.ContainsResolutions <- true
-            callableDiagnostics.Concat(typeDiagnostics).ToLookup(fst, snd)
+            callableDiagnostics
+                .Concat(typeDiagnostics)
+                .ToLookup(fst, fun (_, (position, diagnostic)) ->
+                    { diagnostic with QsCompilerDiagnostic.Range = position + diagnostic.Range })
```

Comment:
Can you explain this in a few words? Does this mutate diagnostics? (I'm not familiar enough with F#).

Reply:
The second argument of the ToLookup, the `fun` lambda, has the effect of mapping pairs of `position` and `diagnostic` to the same `diagnostic` but with the the `position` added to its Range field.

Reply:
Thanks. The question is - if the diagnostics object is modified or a copy is created. If it is modified, I assume the function can only be called once and process each diagnostics once.

Reply:
Ah I understand. A new diagnostic is created. The `{ foo with bar }` syntax creates a new object that copies (deeply I think) the fields of the object `foo` and adds new fields or overrides fields for the new object from the expression `bar`. Here is an official reference on the subject: https://fsharpforfunandprofit.com/posts/records/#creating-records-from-other-records

Reply:
Yes, `QsCompilerDiagnostic` is an immutable record so it cannot be modified - it has to be copied. `with` is a shallow copy, but the fields in `QsCompilerDiagnostic` are also immutable, so it should be equivalent to a deep copy.